src/Akka.Streams.Kafka/Stages/Consumers/Abstract/SingleSourceStageLogic.cs

```diff
@@ -1,132 +1,146 @@
                     break;
-                case AssignmentWithOffset awo:
-                    _consumer.Assign(awo.TopicPartitions);
+                case IManualSubscription manualSubscription:
+                    ConfigureManualSubscription(manualSubscription);
                     break;
+                default:
+                    throw new NotSupportedException();
             }
+        }
 
-            _partitionsAssigned = GetAsyncCallback<IEnumerable<TopicPartition>>(PartitionsAssigned);
-            _partitionsRevoked = GetAsyncCallback<IEnumerable<TopicPartitionOffset>>(PartitionsRevoked);
+        /// <inheritdoc />
+        protected override IActorRef CreateConsumerActor()
+        {
+            var partitionsAssignedHandler = GetAsyncCallback<IEnumerable<TopicPartition>>(PartitionsAssigned);
+            var partitionsRevokedHandler = GetAsyncCallback<IEnumerable<TopicPartitionOffset>>(PartitionsRevoked);
+            var errorOccuredHandler = GetAsyncCallback<Error>(HandleConsumeError);
+
+            var eventHandler = new AsyncCallbacksConsumerEventHandler(errorOccuredHandler, partitionsAssignedHandler, partitionsRevokedHandler);
+            
+            if (!(Materializer is ActorMaterializer actorMaterializer))
+                throw new ArgumentException($"Expected {typeof(ActorMaterializer)} but got {Materializer.GetType()}");
+            
+            var extendedActorSystem = actorMaterializer.System.AsInstanceOf<ExtendedActorSystem>();
+            var actor = extendedActorSystem.SystemActorOf(KafkaConsumerActorMetadata.GetProps(SourceActor.Ref, _settings, eventHandler), 
+                                                          $"kafka-consumer-{_actorNumber}");
+            return actor;
         }
 
         public override void PostStop()
         {
-            Log.Debug($"Consumer stopped: {_consumer.Name}");
-            _consumer.Dispose();
+            ConsumerActor.Tell(new KafkaConsumerActorMetadata.Internal.Stop(), SourceActor.Ref);
 
             base.PostStop();
         }
 
-        //
-        // Consumer's events
-        //
+        protected override void PerformShutdown()
+        {
+            SetKeepGoing(true);
+            
+            if (!IsClosed(_shape.Outlet))
+                Complete(_shape.Outlet);
+            
+            SourceActor.Become(ShuttingDownReceive);
+            StopConsumerActor();
+        }
 
-        private void HandlePartitionsAssigned(IConsumer<K, V> consumer, List<TopicPartition> list)
+        private void ShuttingDownReceive(Tuple<IActorRef, object> args)
         {
-            _partitionsAssigned(list);
+            switch (args.Item2)
+            {
+                case Terminated terminated when terminated.ActorRef.Equals(ConsumerActor):
+                    OnShutdown();
+                    CompleteStage();
+                    break;
+                default:
+                    return;
```

Comment:
So we ignore all messages until the `ConsumerActor` is terminated?

Reply:
Yes, this is what alpakka does.
I think this is mostly because we get to this actor state only after getting `onDownstreamFinished` state event, so no more point of using this messages anyway.